particle-iot/device-os
# BLE: user can optionally send characteristic value with/without ACK.

### Problem

For local characteristic with `NOTIFY` and `INDICATE` property, it will only send notification even if peer device subscribed indication.

For peer characteristic with `NOTIFY` and `INDICATE` property, it will only subscribe notification.

For peer characteristic with `WRITE` and `WRITE_WO_RSP` property, it will only send data without response.

### Solution

- Introduces `BleCharacteristic::setValue(..., BleTxRxType)` API for user to optionally send data with/without acknowledgement.
- Subscribes both notification and indication if both properties are presented with peer characteristic.

### Steps to Test

1. Build and flash the below example app to two Gen3 devices respectively.
2. Connect two devices to serial terminal respectively.
3. Simply type '1', '2', and '3' on one side terminal to see if another terminal can successfully receive these characters, and vice versa.

### Example App
BLE Peripheral:
```c
#include "Particle.h"

#define UART_TX_BUF_SIZE        20

SYSTEM_MODE(MANUAL);

void onDataReceived(const uint8_t* data, size_t len, const BlePeerDevice& peer, void* context);

const char* serviceUuid = "6E400001-B5A3-F393-E0A9-E50E24DCCA9E";
const char* rxUuid = "6E400002-B5A3-F393-E0A9-E50E24DCCA9E";
const char* txUuid = "6E400003-B5A3-F393-E0A9-E50E24DCCA9E";

BleCharacteristic txCharacteristic("tx",
                                   BleCharacteristicProperty::NOTIFY | BleCharacteristicProperty::INDICATE,
                                   txUuid,
                                   serviceUuid);

BleCharacteristic rxCharacteristic("rx",
                                   BleCharacteristicProperty::WRITE_WO_RSP | BleCharacteristicProperty::WRITE,
                                   rxUuid,
                                   serviceUuid,
                                   onDataReceived, &rxCharacteristic);

uint8_t txBuf[UART_TX_BUF_SIZE];
size_t txLen = 0;

void onDataReceived(const uint8_t* data, size_t len, const BlePeerDevice& peer, void* context) {
    for (uint8_t i = 0; i < len; i++) {
        Serial.write(data[i]);
    }
}

void setup() {
    Serial.begin(115200);

    LOG(TRACE, "Application started.");

    BLE.addCharacteristic(txCharacteristic);
    BLE.addCharacteristic(rxCharacteristic);

    BleAdvertisingData data;
    data.appendServiceUUID(serviceUuid);
    BLE.advertise(&data);
}

void loop() {
    if (BLE.connected()) {
        while (Serial.available() && txLen < UART_TX_BUF_SIZE) {
            txBuf[txLen++] = Serial.read();
            Serial.write(txBuf[txLen - 1]);
        }

        if (txLen > 0) {
            if (txBuf[0] == '1') {
                txCharacteristic.setValue(txBuf, txLen);
            } else if (txBuf[0] == '2') {
                txCharacteristic.setValue(txBuf, txLen, BleTxRxType::ACK);
            } else {
                txCharacteristic.setValue(txBuf, txLen, BleTxRxType::NACK);
            }
            txLen = 0;
        }
    }
}
```
BLE Central:
```c
#include "Particle.h"

SYSTEM_MODE(MANUAL);

#define UART_TX_BUF_SIZE        20
#define SCAN_RESULT_COUNT       20

BleScanResult results[SCAN_RESULT_COUNT];

BleCharacteristic peerTxCharacteristic;
BleCharacteristic peerRxCharacteristic;
BlePeerDevice peer;

uint8_t txBuf[UART_TX_BUF_SIZE];
size_t txLen = 0;

void onDataReceived(const uint8_t* data, size_t len, const BlePeerDevice& peer, void* context) {
    for (uint8_t i = 0; i < len; i++) {
        Serial.write(data[i]);
    }
}

void setup() {
    Serial.begin(115200);
    peerTxCharacteristic.onDataReceived(onDataReceived, &peerTxCharacteristic);
}

void loop() {
    if (BLE.connected()) {
        while (Serial.available() && txLen < UART_TX_BUF_SIZE) {
            txBuf[txLen++] = Serial.read();
            Serial.write(txBuf[txLen - 1]);
        }
        if (txLen > 0) {
            if (txBuf[0] == '1') {
                peerRxCharacteristic.setValue(txBuf, txLen);
            } else if (txBuf[0] == '2') {
                peerRxCharacteristic.setValue(txBuf, txLen, BleTxRxType::ACK);
            } else {
                peerRxCharacteristic.setValue(txBuf, txLen, BleTxRxType::NACK);
            }
            txLen = 0;
        }
    }
    else {
        delay(3000);
        size_t count = BLE.scan(results, SCAN_RESULT_COUNT);
        if (count > 0) {
            for (uint8_t i = 0; i < count; i++) {
                BleUuid foundServiceUUID;
                size_t svcCount = results[i].advertisingData.serviceUUID(&foundServiceUUID, 1);
                if (svcCount > 0 && foundServiceUUID == "6E400001-B5A3-F393-E0A9-E50E24DCCA9E") {
                    peer = BLE.connect(results[i].address);
                    if (peer.connected()) {
                        peer.getCharacteristicByDescription(peerTxCharacteristic, "tx");
                        peer.getCharacteristicByUUID(peerRxCharacteristic, "6E400002-B5A3-F393-E0A9-E50E24DCCA9E");
                    }
                    break;
                }
            }
        }
    }
}
```

### References

N/A
